# Week 4
### 张晰 517070910022

## 1
1. Sometimes the loss of significance error can be avoided by rearranging terms in the function using a known identity from trigonometry or algebra. Find an equivalent formula for the following functions that avoids a loss of significance.

    a) $\ln(x+1)-\ln(x)$ for large $x$

    b) $\sqrt{x^2+1}-x$ for large $x$

    c) $\cos^2(x)-\sin^2(x)$ for $x\approx \pi/4$

    d) $\sqrt{\dfrac{1+\cos(x)}{2}}$ for $x\approx \pi$

**Solution:**  
    a) $$\ln(x+1)-\ln(x) = \ln(\frac{x+1}{x}) = \ln(1+\frac{1}{x})$$
    b) $$\sqrt{x^2+1}-x = x\sqrt{1+\frac{1}{x^2}}-x = x\left(\sqrt{1+\frac{1}{x^2}}-1\right)$$
    c) $$ \cos^2(x) - \sin^2(x) = \cos(2x)$$
    d) $$\sqrt{\frac{1+\cos(x)}{2}} = \sqrt{\frac{1+2\cos^2(x/2)-1}{2}} = \left|\cos\left(\frac{x}2\right)\right|$$

## 2
Suppose the algorithmic error is inversely proportional to the square of the number of steps $N$, that is $\epsilon_{\text{approx}}\simeq\cfrac{1}{N^{2}}%$ Estimate the "best approximation" for a double-precision calculation and the corresponding value of $N$. Repeat the calculation in case when the algorithmic error decreases more rapidly as $\epsilon_{\text{approx}}\simeq\cfrac{2}{N^{4}}%$

**Solution:**  
    For double-percision, $\epsilon_m\approx 1e-16$. If $\epsilon_{\rm{approx}}\simeq \frac{1}{N^2}$
    $$\epsilon_{tot}\simeq \frac{1}{N^2} + \sqrt{N}\epsilon_m$$
    When $$\frac{1}{N^2} = \sqrt{N}\epsilon_m$$ 
    It's the "best" approximation. Get $$N \simeq 2.5e6 \qquad \epsilon_{tot}\simeq 3.17e-13$$
    
   If $\epsilon_{\rm{approx}}\simeq \frac{2}{N^2}$,
   When $$\frac{2}{N^4} = \sqrt{N}\epsilon_m$$ 
    It's the "best" approximation. Get $$N \simeq 4200 \qquad \epsilon_{tot}\simeq 1.28e-14$$

## 3
The value of $\pi$ can be estimated from the calculations of the side lengths of regular polygons inscribing a circle. In general, $\pi_k=\pi_\infty+\dfrac{c_1}{k}+\dfrac{c_2}{k^2}+\dfrac{c_3}{k^3}+\cdots$ where $\pi_k$ is the ratio of the perimeter to the diameter of a regular $k$-sided polygon. Determine the approximate value of $\pi\approx\pi_\infty$ from $\pi_8 = 3.061 467, \pi_{16} = 3.121 445$,  $\pi_{32} = 3.136 548$, and $\pi_{64} = 3.140 331$ of the inscribing polygons. Which $c_i$ is most significant and why?


**Solution:**  
We have 4 points on the equation, neglect the terms over $\left(\frac{1}{k}\right)^3$, we can get a equation set:
$$
\left\{\begin{aligned}
3.061467 &= \pi_\infty + \frac{c_1}{8} + \frac{c_2}{8^2} + \frac{c_3}{8^3}\\
3.121445 &= \pi_\infty + \frac{c_1}{16} + \frac{c_2}{16^2} + \frac{c_3}{16^3}\\
3.136548 &= \pi_\infty + \frac{c_1}{32} + \frac{c_2}{32^2} + \frac{c_3}{32^3}\\
3.140331 &= \pi_\infty + \frac{c_1}{64} + \frac{c_2}{64^2} + \frac{c_3}{64^3}
\end{aligned}\right.
$$
Solve this equation set with numpy:

In [1]:
import numpy as np
matrix = np.array([[1,1/8,1/8**2,1/8**3],[1,1/16,1/16**2,1/16**3],[1,1/32,1/32**2,1/32**3],[1,1/64,1/64**2,1/64**3]])
value = np.array([[3.061467],[3.121445],[3.136548],[3.140331]])
coff = np.linalg.solve(matrix,value)
print(f'pi = {coff[0,0]}')

pi = 3.1415837619047626


We can see $\pi_{\infty}\approx 3.14158376$.
Of course $c_1$ is the most important. As $k$ rising up, $\left(\frac{1}{k}\right)^n$ with $n\geq 2$ are all much smaller than $\frac{c_1}{k}$

## 4
   Consider the fixed-point iteration described by
    $$
    x_{n+1}= g( x_n)=\frac{1}{2}\left(x_n+\frac{a}{x_n}\right)
    $$
Apply this iteration function $g(x)$ to approximate $\sqrt{5}$.
    
Hint: Please write your own script based on the sample scripts that implement the bisection method/secant method. 


**Solution:**  
With $$ x^* = \frac{1}{2}(x^* + \frac{a}{x^*})\\\downarrow$$
$$x^* = \sqrt{a}$$
So to get $\sqrt{5}$, we should choose $ a = 5 $, and solve the equation 
$$x = \frac{1}{2}\left(x+\frac{5}{x}\right)$$

In [2]:
def Fixed_Point_Method(f,start_point,N=200,precision = 1e-10):
    count = 0
    while count < N:
        count += 1
        if abs(start_point-f(start_point) / start_point) < precision:
            break
            
        start_point = f(start_point)
    return start_point

print(Fixed_Point_Method(lambda x:1/2*(x+5/x),3))

2.23606797749979


In [3]:
#二分法，递归
def Bisection_Method(f,left,right,precision=1e-16):
    if f(left)==0:
        return left
    elif f(right) == 0 :
        return right
    elif abs(right - left)< precision:
        return (left + right) / 2
    elif f(left) * f((left + right) / 2) > 0:       # 若f(left)和f( (left+right)/2 )同号，就用(left+right)/2代替left
        return Bisection_Method(f,(left + right) / 2,right, precision)
    else :
        return Bisection_Method(f,left , (left + right) / 2, precision)
    
print(Bisection_Method(lambda x:1/2*(5/x - x),1,3)) #left别取0，没写try

2.23606797749979


In [4]:
def Secant_Method(f,x1,x2,precision=1e-19):
    if f(x1) == 0:
        return x1
    elif f(x2) == 0 :
        return x2
    elif abs(x1 - x2) < precision:
        return (x1 + x2) / 2
    else :
        x3 = x2 - f(x2) * ( x2 - x1 ) / ( f(x2) - f(x1) )
        return Secant_Method(f,x2,x3)
        
print(Secant_Method(lambda x:1/2*(5/x - x),1,3))

2.23606797749979
